framework/src/modules/interoperability/mainchain/commands/cc_update.ts

```diff
@@ -12,18 +12,433 @@
+			};
+		}
+		if (
+			txParams.activeValidatorsUpdate &&
+			txParams.activeValidatorsUpdate.length !== 0 &&
+			txParams.newCertificateThreshold > BigInt(0)
+		) {
+			// Non-empty certificate
+			if (!txParams.certificate) {
+				return {
+					status: VerifyStatus.FAIL,
+					error: new Error('Certificate cannot be empty.'),
+				};
+			}
+
+			// params.activeValidatorsUpdate has the correct format
+			for (let i = 0; i < txParams.activeValidatorsUpdate.length; i += 1) {
+				const currentValidator = txParams.activeValidatorsUpdate[i];
+				const nextValidator = txParams.activeValidatorsUpdate[i + 1];
+				if (currentValidator.blsKey.byteLength !== VALID_BLS_KEY_LENGTH) {
+					return {
+						status: VerifyStatus.FAIL,
+						error: new Error(`BlsKey length should be equal to ${VALID_BLS_KEY_LENGTH}.`),
+					};
+				}
+				if (currentValidator.blsKey.compare(nextValidator.blsKey) > -1) {
+					return {
+						status: VerifyStatus.FAIL,
+						error: new Error('Validators blsKeys must be unique and lexicographically ordered'),
+					};
+				}
+			}
+		}
+
+		const partnerChannelStore = context.getStore(transaction.moduleID, STORE_PREFIX_CHANNEL_DATA);
+		const partnerChannelData = await partnerChannelStore.getWithSchema<ChannelData>(
+			partnerChainIDBuffer,
+			channelSchema,
+		);
+		// InboxUpdate Validity
+		const { crossChainMessages, messageWitness, outboxRootWitness } = txParams.inboxUpdate;
+		const ccmHashes = crossChainMessages.map(ccm => hash(ccm));
+
+		let newInboxRoot;
+		let newInboxAppendPath;
+		let newInboxSize;
+		for (const ccm of ccmHashes) {
+			const { appendPath, size } = regularMerkleTree.calculateMerkleRoot({
+				value: ccm,
+				appendPath: partnerChannelData.inbox.appendPath,
+				size: partnerChannelData.inbox.size,
+			});
+			newInboxAppendPath = appendPath;
+			newInboxSize = size;
+		}
+		if (txParams.certificate) {
+			// If inboxUpdate contains a non-empty messageWitness, then update newInboxRoot to the output
+			newInboxRoot = regularMerkleTree.calculateRootFromRightWitness(
+				newInboxSize as number,
+				newInboxAppendPath as Buffer[],
+				messageWitness.siblingHashes,
+			);
+
+			const proof = {
+				siblingHashes: outboxRootWitness.siblingHashes,
+				queries: [
+					{
+						key: partnerChainIDBuffer,
+						value: newInboxRoot,
+						bitmap: outboxRootWitness.bitmap,
+					},
+				],
+			};
+			const outboxKey = rawStateStoreKey(STORE_PREFIX_OUTBOX_ROOT);
+			const querykeys = [outboxKey];
+			const isSMTRootValid = sparseMerkleTree.verify(
+				querykeys,
+				proof,
+				decodedCertificate.stateRoot,
+				SMT_KEY_LENGTH,
+			);
+			if (!isSMTRootValid) {
+				return {
+					status: VerifyStatus.FAIL,
+					error: new Error(
+						'Failed at verifying state root when messageWitness and certificate are non-empty.',
+					),
+				};
+			}
+		} else if (!txParams.certificate) {
+			newInboxRoot = regularMerkleTree.calculateRootFromRightWitness(
```

Comment:
here also check of `// If inboxUpdate contains a non-empty messageWitness` is missing